### PR TITLE
Line wrap for long rows (w toggle + wrap_lines config)

### LINKED ISSUE #31: wrap lines on pane width limit (line wrap)
**Problem**

Too long lines are truncated;

Image

So I cannot see the rest of the line/code; 

**Proposed behavior**

Let the line "truncation" behavior be configurable, a line wrap feature;

Something like that:
Image

**Alternatives considered**

...

### PR DESCRIPTION
**Pull request by JonasBaeumer**
Closes #31

Long lines currently clip at the pane edge and can only be panned horizontally, which is easy to miss and awkward for prose-heavy files. This adds a wrap mode:

- **`w`** in the diff/source pane toggles between clip-and-pan (unchanged default) and wrapping long rows to the pane width. While wrapped, the pan keys and horizontal wheel are inert; `0`/pan come back when toggled off.
- **`wrap_lines`** config key (default `false`) starts every review wrapped for reviewers who prefer it. The pane now reads the plugin config itself; the wire protocol is untouched.

## How it works

Wrapped continuation lines are display-map extras, exactly the mechanism inline comment rows already use: they hang under their base row in `DispMap`, so cursor follow (`follow_display` already keeps a row's extras visible), mouse hits (a click on a continuation resolves to its logical line), folds, goto, and comment weaving all compose with no wrap-specific cases. The splitter shares `pan_and_clip`'s two invariants — unicode-width 0.2 display widths and UAX #29 grapheme-cluster atomicity (a straddling cluster moves to the next line whole). Continuations sit behind a gutter-width indent so code columns stay aligned; hunk headers wrap across the full width; fold pills never wrap (the display map skips them the same way).

## Testing

8 new tests (160 total), including a `TestBackend` render-parity test pinning `draw_diff`'s expansion to `disp_map`'s row math — the one place a drift would silently misplace scroll offsets and clicks. Splitter edge cases covered: fitting rows untouched, indent width, wide-cluster straddling, single-span rows, folded rows contributing no extras, pan keys silenced while wrapped.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by JonasBaeumer** [comment]
@codex review

File: src/pane.rs [hunk 1]
```diff
@@ -28,7 +28,10 @@ pub fn run() -> Result<()> {
     let (mut channel, goto_rx) = conn.into_channel();
 
     let model = crate::diff::load(&request.working_dir, request.baseline.as_deref());
-    let outcome = crate::ui::run(&request, model, goto_rx)?;
+    // The pane reads the same config file as the MCP server; only the
+    // display keys matter here (placement/timeouts are the server's side).
+    let config = crate::config::load();
+    let outcome = crate::ui::run(&request, model, goto_rx, config.wrap_lines)?;
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/pane.rs:34]: Confirmed: the pane process calls `config::load()` at open (src/pane.rs), so `wrap_lines` picks up edits per pane while placement/timeouts stay on the server snapshot. Behavior is intentional (the pane owns display keys, per the comment at that call site), so the fix documents the exception: docs/configuration.md now states `wrap_lines` is read when the pane opens and applies from the next review. Fixed in bd3e03b.

File: src/ui.rs [hunk 2]
```diff
@@ -3243,6 +3324,82 @@ fn pan_and_clip(line: &mut Line<'static>, hscroll: usize, width: usize, pinned:
...
+    let mut out = Vec::with_capacity(chunks.len());
+    for (i, chunk) in chunks.into_iter().enumerate() {
+        let mut spans: Vec<Span<'static>> = Vec::new();
+        if i == 0 {
+            spans.extend(line.spans.iter().take(pinned).cloned());
+        } else if pinned > 0 {
+            spans.push(Span::styled(" ".repeat(pinned_cols), indent_style));
+        }
+        spans.extend(chunk);
+        out.push(Line::from(spans));
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Verified: `Line::styled` puts the color on `Line::style` over a default-styled span, and `Line::from(spans)` in the reconstruction path reset it, so long hunk headers and placeholders lost their color exactly when wrapped (the fits-on-one-line fast path clones and keeps it). Fixed in 00cd09c by applying `line.style` to every generated row; regression test `wrap_line_keeps_the_line_level_style` fails without the fix, and the full suite (161) passes with it.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 3]
```diff
@@ -1932,6 +1942,26 @@ impl<'a> App<'a> {
...
+        // Rows inside a fold run contribute nothing — the tail isn't on
+        // screen, and the head is replaced by its (never-wrapped) pill.
+        if self.wrap {
+            let rows: Option<&Vec<Line<'static>>> = match self.view {
+                ViewMode::Diff => self.row_cache.get(&self.nav.selected),
+                ViewMode::Source => self
+                    .source_cache
+                    .get(&self.nav.selected)
+                    .and_then(|r| r.as_ref().ok())
+                    .map(|s| &s.lines),
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs]: Verified: `draw_diff` wraps every non-pill visible row including the placeholder, while `disp_map` counted extras only for `Some(Ok(source))`, so the map reported 1 row against 5 rendered for a long reason at width 20. Fixed in 4d9b1e0: the wrap branch now mirrors `view_lines` (error and not-loaded placeholders both counted). Regression test `display_map_counts_wrapped_source_error_placeholder_rows` fails without the fix; full suite (162) passes with it.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 4]
```diff
@@ -2117,14 +2157,14 @@ impl<'a> App<'a> {
             // Up(Left) is handled unconditionally at the top of this
             // function, before the modal gates.
             MouseEventKind::ScrollRight => {
-                if in_diff {
+                if in_diff && !self.wrap {
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:2167]: Verified: the keyboard pan row appends the "inactive while wrapped" note via `pan_note`, while the Mouse section row and docs/controls.md left the horizontal wheel unqualified. Fixed in 2465747: the mouse help row now uses the same `pan_note`, and controls.md says panning (keys and horizontal wheel) is inactive while wrapped. Regression test `help_overlay_notes_horizontal_wheel_inactive_while_wrapped` fails without the fix; full suite (163) passes.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 5]
```diff
@@ -3080,6 +3141,29 @@ fn draw_diff(frame: &mut Frame, area: Rect, app: &App, file: Option<&FileDiff>)
...
+    let mut expand_map: Vec<usize> = Vec::new();
+    if app.wrap {
+        let pills: HashSet<usize> = folds.iter().map(|&(s, _)| folded_row(s)).collect();
+        let mut wrapped: Vec<Line> = Vec::with_capacity(lines.len());
+        for (i, line) in lines.into_iter().enumerate() {
+            expand_map.push(wrapped.len());
+            if pills.contains(&i) {
+                wrapped.push(line);
+            } else {
+                wrapped.extend(wrap_line(&line, inner_width));
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:3167]: Verified, and it is worse than described: past 65,535 rows the clamp does not just strand the cursor, ratatui Paragraph internal u16 math overflows (debug panic at paragraph.rs:483). Fixed in b154e23 by rendering the viewport slice (skip(scroll).take(viewport)) with no Paragraph::scroll at all, so offsets stay usize end to end; scroll_row_u16 and its clamp test are removed. Regression test `rendering_reaches_wrapped_display_rows_past_the_u16_scroll_range` builds ~67k wrapped display rows, scrolls past u16::MAX, and asserts the cursor row renders; it fails without the fix. Full suite (163) passes.

File: src/ui.rs [hunk 6]
```diff
@@ -2389,15 +2429,26 @@ impl<'a> App<'a> {
...
                             self.next_pan_stop(self.diff.hscroll, self.pan_step(term_size))
                                 .min(self.pan_cap())
                     }
-                    KeyCode::Left | KeyCode::Char('H') => {
+                    KeyCode::Left | KeyCode::Char('H') if !self.wrap => {
                         self.diff.hscroll = self.diff.hscroll.saturating_sub(self.pan_step(term_size))
                     }
                     KeyCode::Char('0') => self.diff.hscroll = 0,
+                    KeyCode::Char('w') => {
+                        self.wrap = !self.wrap;
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:2453]: Rebutting: d/u paging in base rows is pre-existing behavior this PR does not touch (page_down/page_up/half_page are unchanged from main), and it already behaves exactly this way when woven comment rows occupy display space. The PR deliberately makes wrap continuations the same DispMap extras as comment rows so movement math has no wrap-specific cases; making d/u display-space aware would either add such a case or change pre-existing paging behavior outside this PR. No content is skipped: ensure_cursor_visible keeps the cursor on screen after every page, and j/k and the wheel reach every display row. If display-space paging is wanted, that is a design call for the lead, and it applies to comment extras on main as much as to wrap.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:3167]: Rebutting with measurements (release build, 2 MiB single-line file, the worst case the caps admit). Per idle frame: wrap_line at a 40-col pane = 66 ms building 72,316 rows; the pre-existing unwrapped path pays pan_and_clip = 46 ms on the same input, every frame, on main today. Wrap is the same cost class as the accepted clip-and-pan draw (~1.4x), not a new freeze: both walk the 2M-grapheme line per tick, and virtualizing wrap would leave the identical per-frame cost in clip mode untouched. The "two million Line/Span objects" figure only occurs at a 12-col pane (one content column after the 11-col gutter), a degenerate layout, and even there it is 249 ms per frame, not exhaustion. Per-frame O(file) draw over the capped sizes is the existing accepted design (view_lines clones every row per frame too); if that budget should change, it is a design call for the lead and applies to clip mode as much as wrap.

**Comment by JonasBaeumer** [comment]
@codex review

File: src/ui.rs [hunk 7]
```diff
@@ -1932,6 +1942,36 @@ impl<'a> App<'a> {
...
+                        placeholder = vec![source_placeholder_line("not loaded")];
+                        Some(&placeholder)
+                    }
+                },
+            };
+            for (b, row) in rows.into_iter().flatten().enumerate() {
+                if fold_run_containing(&folds, b).is_some() {
+                    continue;
+                }
+                ends.extend(std::iter::repeat(b).take(wrap_height(row, inner_width) - 1));
```
[the remark this answers is not in the document]
Comment (JonasBaeumer) [reply src/ui.rs:1979]: Verified and measured: with 40k base rows and one continuation per row, three base_at lookups took ~17s (base_count x |ends| comparisons), and base_at also runs on every wheel scroll, not just clicks. Fixed in a230c11: disp counts comment rows via partition_point on the sorted ends, and base_at binary-searches the monotone disp (rows in a hidden tail plateau on their pill, so monotonicity holds). Regression test `base_at_stays_fast_with_wrap_scale_extras` fails without the fix (17s vs a 1s bound) and passes in microseconds with it; the existing fold/comment DispMap tests pin the lookup values unchanged. Full suite (164) passes.

**Comment by JonasBaeumer** [comment]
@codex review
